Add type annotations and drop Python 3.5 host support

YannickJadoul opened the pull request:
Finally pushing this, after this came up again in #307. As noted, this is based on an old version of master, before a bunch of refactoring PRs were merged, so it still needs to be rebased (which might be as hard as just doing it from scratch again?).

To check, install mypy and run `mypy --strict cibuildwheel/`.

Mainly meant to open the discussion of which aspects of type annoations are useful in the context of cibuildwheel and which are probably not. Python's typing is not all or nothing, so we can surely find middle ground.

@joerick, @Czaki, I think you were both interested to see this? :-)

EDIT:
Closes #258

Czaki commented on cibuildwheel/bashlex_eval.py:
```diff
@@ -37,11 +41,11 @@ def evaluate_node(node, context):
     word_end = node.pos[1]
     word_string = context.input[word_start:word_end]
-    letters = list(word_string)
+    letters = list(word_string)  # type: List[Optional[str]]
```
why optional?

YannickJadoul replied:
Because I got an error that `None` is assigned further down. If you remove the annotation, it should show up.

joerick replied:
We could set them to `''` empty string instead of None, and make it non-optional.

YannickJadoul replied:
If you prefer? It does clean up `''.join(l for l in letters if l is not None)` to just `''.join(letters)` as well.

Czaki commented on cibuildwheel/environment.py:
```diff
@@ -74,8 +70,14 @@ def as_dictionary(self, prev_environment):
         return [a.as_shell_assignment() for a in self.assignments]
 
-    def __repr__(self):
+    def __repr__(self) -> str:
```
I do not think that typing `__repr__` this is needed

YannickJadoul replied:
It probably gave me warnings if I didn't do it. But I'll try.

YannickJadoul replied:
`cibuildwheel/environment.py:56: error: Function is missing a return type annotation`

joerick replied:
I noticed above that some functions were marked as `-> None`, which seemed... too obvious and cluttery. As does this. Maybe we can get a bit more granular with the mypy rules - `--strict` just enables them all

YannickJadoul replied:
`mypy.ini` or `setup.cfg` should allow for that, indeed.

joerick replied:
It looks like the function types really are required, if we want type inference-

> Type inference is not used in dynamically typed functions (those without a function type annotation) — every local variable type defaults to Any in such functions. Any is discussed later in more detail.

https://mypy.readthedocs.io/en/stable/type_inference_and_annotations.html#type-inference

YannickJadoul replied:
Yuk :-(
Somehow it isn't necessary for '__init__`, I think?

Czaki commented on cibuildwheel/linux.py:
```diff
@@ -6,14 +6,21 @@
 
+from typing import Callable, Dict, List, Optional
+
+from .environment import (
+    ParsedEnvironment,
+)
```
single line?

```python
from .environment import ParsedEnvironment,
```

YannickJadoul replied:
Maybe, yes. I guess I copied this, and removed an unused import afterwards.

YannickJadoul replied:
Isn't it nicer in style to have both imports the same?

```python
from .environment import (
    ParsedEnvironment,
)
from .util import (
    download,
    get_build_verbosity_extra_flags,
    prepare_command,
)
```

Czaki replied:
and line above?
```python
 from typing import Callable, Dict, List, Optional
```

Czaki replied:
Yes. this dicusssion is not strongly connected with PR clue.

YannickJadoul replied:
Nono, it's a valid point; I appreciate the comment! :-)
But I don't really know what's the best way to solve it.

Czaki commented on cibuildwheel/linux.py:
```diff
@@ -6,14 +6,21 @@
 
-def get_python_configurations(build_selector):
-    PythonConfiguration = namedtuple('PythonConfiguration', ['identifier', 'path'])
+PythonConfiguration = namedtuple('PythonConfiguration', ['identifier', 'path'])
```
```python
 PythonConfiguration = NamedTuple('PythonConfiguration', [('identifier', str), ('path', str)])
```

Czaki commented on cibuildwheel/macos.py:
```diff
@@ -7,14 +7,19 @@
 
+from typing import Callable, Dict, List, Optional, Union
+
+from .environment import (
+    ParsedEnvironment,
+)
```
```python
from .environment import ParsedEnvironment,
```

Czaki commented on cibuildwheel/windows.py:
```diff
@@ -7,6 +7,11 @@
 
+from typing import Callable, Dict, List, Optional
+
+from .environment import (
+    ParsedEnvironment,
+)
```
```python
from .environment import ParsedEnvironment,
```

Czaki commented on cibuildwheel/macos.py:
```diff
@@ -24,8 +29,10 @@ def call(args, env=None, cwd=None, shell=False):
 
-def get_python_configurations(build_selector):
-    PythonConfiguration = namedtuple('PythonConfiguration', ['version', 'identifier', 'url'])
+PythonConfiguration = namedtuple('PythonConfiguration', ['version', 'identifier', 'url'])
```
```python
PythonConfiguration = NamedTuple('PythonConfiguration', [('version', str), ('identifier', str), ('url, str)'])
```

Czaki commented on cibuildwheel/util.py:
```diff
@@ -24,37 +26,37 @@ def get_build_verbosity_extra_flags(level):
         return match_any(self.build_patterns) and not match_any(self.skip_patterns)
 
-    def __repr__(self):
+    def __repr__(self) -> str:
```
unnecessary

YannickJadoul replied:
Seems necessary to shut up `mypy`? See above.

Czaki commented on cibuildwheel/windows.py:
```diff
@@ -31,22 +36,24 @@ def simple_shell(args, env=None, cwd=None):
 
-def get_python_configurations(build_selector):
-    PythonConfiguration = namedtuple('PythonConfiguration', ['version', 'arch', 'identifier', 'url'])
+PythonConfiguration = namedtuple('PythonConfiguration', ['version', 'arch', 'identifier', 'url'])
```
```python
PythonConfiguration = NamedTuple('PythonConfiguration', [('version', str), ('arch', str), ('identifier', str), ('url', Optional[str])])
```

YannickJadoul commented:
> Great job.

Thanks! Thanks for the quick review as well.

There's many things I'm not convinced of that I like, though:



> Looking on this code I think that `installl_cpython` and `install_pypy` should get as argument `PythonConfiguration`, not only selected fields. This will give better flexibility.

I don't know if we need that flexibility currently, but I'll have a look after rebasing. Too many things have changed anyway.

> And when drop execution on python 3.5 they maybe it is a good idea to return `Path` object from them.

There are lots of things to do when dropping 3.5 support. `pathlib` is not really related to types, I'd say (though types will help refactoring). For types, most importantly, there's `var: Type = abc` rather than `var = abc  # type: Type` and the NamedTuple. I'll make an issue with a list!

YannickJadoul commented:
There are still quite a few things I don't like, though:

- I don't mind importing `typing` for type annotations, but it's weird/ugly that it suddenly takes over the role of `collections.namedtuple`. `typing` is for types, not for functionality!?
- The `@overload`s of `__main__.get_option_from_environment` are exaggerated (I couldn't find another way to specify the return type correctly than with `@overload`). There's more code to define/type the function than in it's body. And flake8 acts annoying about style, as it wants me to put `...` on the next line ánd have 2 lines of whitespace in between (so I added `# noqa`). Kind of defeats the purpose of a quick utility function?
- Importing `ParsedEnvironment` is quite annoying as well? Where's the nice duck-typing gone?

Czaki commented:
> There are still quite a few things I don't like, though:
> 
>    * I don't mind importing `typing` for type annotations, but it's weird/ugly that it suddenly takes over the role of `collections.namedtuple`. `typing` is for types, not for functionality!?

For me best is dataclass, but it is introduced in python 3.7. Current option is only one which gives type support and do not force us to write own constructor. 

>    * The `@overload`s of `__main__.get_option_from_environment` are exaggerated (I couldn't find another way to specify the return type correctly than with `@overload`). There's more code to define/type the function than in it's body. And flake8 acts annoying about style, as it wants me to put `...` on the next line ánd have 2 lines of whitespace in between (so I added `# noqa`). Kind of defeats the purpose of a quick utility function?

maybe use `pyi` for some files (it need to move all typing to this file) to hide typing from main file. 

>    * Importing `ParsedEnvironment` just for typing is quite annoying as well? Where's the nice duck-typing gone?

Something for something. You got type cheeking, but need to write more code. 

You do not think about introduce type check task (as flake8) to tests?

YannickJadoul commented:
> You do not think about introduce type check task (as flake8) to tests?

Yes, I will. But figuring that out was not my main concern. Something to think about after rebase, and once we can make decisions.
That's why I just wrote the `mypy --strict cibuildwheel/` command, for now.

joerick commented on cibuildwheel/__main__.py:
```diff
@@ -14,11 +16,15 @@
+@overload
+def get_option_from_environment(option_name: str, platform: Optional[str], default: str) -> str: ...  # noqa: E704
+@overload
+def get_option_from_environment(option_name: str, platform: Optional[str] = None, default: None = None) -> Optional[str]: ...  # noqa: E704 E302
```
Is this overload redundant? It feels like the one below captures this one

YannickJadoul replied:
The problem then is that `mypy` gets confused that even if you pass a default (and the return value can thus never be `None`), it will complain about assigning an `Optional[str]` to a `str` (e.g, calling `parse_environment(environment_config)` which expects a `str`).

joerick replied:
yes, I meant that specific line - I'll try it

joerick replied:
(you probably already know this, but) The way that `@overload` works is really confusing... the signatures marked with @overload are the 'public' signatures of the function, but the last one specifies the types of the arguments inside. So this is equivalent:

```python
@overload
def get_option_from_environment(option_name: str, platform: Optional[str], default: str) -> str: ...  # noqa: E704
@overload
def get_option_from_environment(option_name: str, platform: Optional[str] = None, default: Optional[str] = None) -> Optional[str]: ...  # noqa: E704 E302
def get_option_from_environment(option_name: str, platform: Optional[str] = None, default: Optional[str] = None) -> Optional[str]:  # noqa: E302
    '''
    Returns an option from the environment, optionally scoped by the platform.
```

According to the docs, this is the intended way to use `@overload`:

```python
@overload
def get_option_from_environment(option_name: str, platform: Optional[str], default: str) -> str: ...  # noqa: E704
@overload
def get_option_from_environment(option_name: str, platform: Optional[str] = None, default: Optional[str] = None) -> Optional[str]: ...  # noqa: E704 E302
def get_option_from_environment(option_name, platform=None, default=None):
    '''
    Returns an option from the environment, optionally scoped by the platform.
```

But in this form, `--strict` complains `Function is missing a type annotation`. Without `--strict`, mypy has those params as `Any`, so isn't type-checking the interior. Frustrating.

YannickJadoul replied:
Yes, I tried lots of stuff, when writing that. Also with `Union[]` and generics, but I didn't get anything to work (and pass with `--strict`), so best I could do was what's there.
Very frustrating for such a simple use case, indeed :-(

joerick commented on cibuildwheel/windows.py:
```diff
@@ -101,7 +106,10 @@ def install_pypy(version, arch, url):
 
 
-def build(project_dir, output_dir, test_command, test_requires, test_extras, before_build, build_verbosity, build_selector, repair_command, environment):
+def build(project_dir: str, output_dir: str, build_selector: Callable[[str], bool],
```
build_selector should be a BuildSelector object, I think.

YannickJadoul replied:
If you prefer, yes. I made it this way, such the code here (and in `macos.py` and `linux.py`) is agnostic about the actual type (and doesn't need to import it), but just knows it need to be able to be called.

YannickJadoul commented:
Is there any PR I should wait for before rebasing? If not, I'll try to do it somewhere over the next few days. I've noticed it doesn't make a lot of sense to discuss code too much, if it might still change substantially.

joerick commented:
So. In general, I'm happy with the way this is heading. Yes, a detailed review would have to wait until we have an up-to-date PR. Let's try to get a few of the open PRs in before doing that - I'm thinking about #295 and #311 mainly, but some of the smaller ones might as well land, too.

joerick commented:
> * I don't mind importing `typing` for type annotations, but it's weird/ugly that it suddenly takes over the role of `collections.namedtuple`. `typing` is for types, not for functionality!?

Strong agree. Always felt weird to me. Why didn't they just type-enable the version in `collections`? But it's not a big deal to me.

> * The `@overload`s of `__main__.get_option_from_environment` are exaggerated (I couldn't find another way to specify the return type correctly than with `@overload`). There's more code to define/type the function than in it's body. And flake8 acts annoying about style, as it wants me to put `...` on the next line ánd have 2 lines of whitespace in between (so I added `# noqa`). Kind of defeats the purpose of a quick utility function?

Yeah. I don't think there's anything we can do to mypy to improve this, but maybe we can tweak the flake8 rules to help here. Let's explore in the new/rebased PR?

> * Importing `ParsedEnvironment` just for typing is quite annoying as well? Where's the nice duck-typing gone?

Will presumably be helped by the BuildOptions struct in #311.

joerick commented:
Another note - I vote we drop Python 3.5 support in the same PR that we do the typing work. It will make things much easier. But let's leave the `pathlib` refactor for now - that will be much easier once we have types.

YannickJadoul commented:
Side note, after all the complaining: on the bright side, running `mypy` and annotating and fixing these issues is a lot of fun and satisfaction as well ;-) (I just wish there'd be less clutter remaining in the code afterwards.)

YannickJadoul commented:
First rebase. Got curious, and it actually went a lot quicker than expected. Juts a few extra things to annotate, now.

Czaki commented on cibuildwheel/util.py:
```diff
@@ -2,12 +2,13 @@
 
 
-def prepare_command(command, **kwargs):
+def prepare_command(command: str, **kwargs: str) -> str:
```
str? `Dict`

YannickJadoul replied:
No: https://stackoverflow.com/questions/37031928/type-annotations-for-args-and-kwargs

joerick commented on cibuildwheel/util.py:
```diff
@@ -81,18 +82,21 @@ def download(url, dest):
 
     @classmethod
-    def with_defaults(cls):
+    def with_defaults(cls: Type[DependencyConstraints_T]) -> DependencyConstraints_T:
```
We don't have to worry about subclassing, maybe lets just make this a staticmethod? Or do-

```suggestion
    def with_defaults(cls: Type[DependencyConstraints]) -> DependencyConstraints:
```

YannickJadoul replied:
OK, let's go for `@staticmethod`, then?
I know I overdid it a little bit, here. But just wanted do get `--strict` to pass, and worry about what was too much later :-)

joerick replied:
Yep staticmethod is good

YannickJadoul replied:
Should be done, already (but GitHub shows an old version if you click the `cibuildwheel/util.py` here).

Czaki commented on cibuildwheel/windows.py:
```diff
@@ -62,20 +70,21 @@ def get_python_configurations(build_selector):
-def install_pypy(version, arch, url):
+def install_pypy(version: str, arch: str, url: Optional[str]) -> str:
     assert arch == '32'
+    assert url is not None
```
Optional and assert? Maybe add to if before call function?

YannickJadoul replied:
I didn't want to break the line `installation_bin_path = install_pypy(python_configuration.version, python_configuration.url)`. But yeah, maybe.

Czaki replied:
I think about ` elif python_configuration.identifier.startswith('pp') and python_configuration.url is not None:

YannickJadoul replied:
Yeah, just did that. Just give me a second to commit :-)

YannickJadoul replied:
Oh, wait no. I did it slightly differently.

But if `python_configuration.identifier.startswith('pp')` and `python_configuration.url is None`, that's a mistake. So we should assert on that.

Czaki replied:
so tests will fail on unsuported python interpreter
Or I have maybe introduce CPythonConfiguration and PyPythonConfiguration? To force at the level of code url?

YannickJadoul replied:
What? No. `python_configuration.url` is something we list ourselves. How can this fail?

YannickJadoul replied:
This is literally what asserts are made for, right? To check programmatic assumptions that are not user input.

YannickJadoul replied:
(Or I'm really missing something, and then I'm curious what obvious thing I overlooked. Sorry, then, for being too unbelieving :-) )

YannickJadoul commented on CI.md:
```diff
@@ -1,11 +1,11 @@
-| Linux    | Travis CI        | CircleCI         | AppVeyor / GitHub Actions                         | Azure Pipelines  |
-| macOS    | Azure Pipelines  | CircleCI         | AppVeyor / Travis CI¹ / CircleCI / GitHub Actions | Azure Pipelines  |
-| Windows  | TravisCI         | Azure Pipelines  | AppVeyor / GitHub Actions                         | Azure Pipelines  |
+|          | 3.6                          | 3.7                                               | 3.8              |
+|----------|------------------------------|---------------------------------------------------|------------------|
+| Linux    | Travis CI / CircleCI         | AppVeyor / GitHub Actions                         | Azure Pipelines  |
+| macOS    | CircleCI                     | AppVeyor / Travis CI¹ / CircleCI / GitHub Actions  | Azure Pipelines  |
+| Windows  | Travis CI / Azure Pipelines  | AppVeyor / GitHub Actions                         | Azure Pipelines  |
```
I added it in #265 (see https://github.com/joerick/cibuildwheel/pull/265#issuecomment-581493314 and https://github.com/joerick/cibuildwheel/pull/265#issuecomment-582149255) and and I think it was updated after GitHub Actions was added.

It's not part of the docs and not meant for users, but just something I added to keep track of which CI providers our own tests run on and which Python versions are run (after @mayeut made such a nice table once or twice at the start of a PR; sorry for the useless mention, @mayeut, but it was a great idea :-) ).

YannickJadoul commented on cibuildwheel/util.py:
```diff
@@ -81,18 +82,21 @@ def download(url, dest):
 
     @classmethod
-    def with_defaults(cls):
+    def with_defaults(cls: Type[DependencyConstraints_T]) -> DependencyConstraints_T:
```
Because `cls` is a parameter and could be different if there's a subclass. If this isn't necessary, this should probably not be `@classmethod` but `@staticmethod`?

Czaki replied:
I still does not understand. If it is subclass it still return instance of `DependecyConstrains` and cls still be `DependecyConstrains` class (with extension)

YannickJadoul replied:
No, because you expect `SubclassOfDependencyConstraints.with_defaults` to return a `SubclassOfDependencyConstraints` object, and not a `DependencyConstraints` object. So the return statement would be OK, but the code using the classmethod wouldn't be OK.

YannickJadoul replied:
Anyway, it's solved. See https://github.com/joerick/cibuildwheel/pull/315#discussion_r406854654.

Czaki replied:
So still typing system in python have weakness. Declaration that it return something in type of current class should be easy.

YannickJadoul commented on .travis.yml:
```diff
@@ -30,25 +30,25 @@ jobs:
-    - &linux_s390x_35
-      name: Linux | s390x | Python 3.5
+    - &linux_s390x_36
+      name: Linux | s390x | Python 365
```
Whoops. Thanks :-)

joerick commented on cibuildwheel/windows.py:
```diff
@@ -107,6 +106,7 @@ def setup_python(python_configuration: PythonConfiguration, dependency_constrain
     if python_configuration.identifier.startswith('cp'):
         installation_path = install_cpython(python_configuration.version, python_configuration.arch, nuget)
     elif python_configuration.identifier.startswith('pp'):
+        assert python_configuration.url is not None
         installation_path = install_pypy(python_configuration.version, python_configuration.arch, python_configuration.url)
```
Really cool that this works!

YannickJadoul replied:
Yes, indeed. Makes perfect sense, but cool feature of mypy type inference :-)

Czaki commented:
Progress?

YannickJadoul commented:
> Progress?

On what? I think there aren't a lot of open discussions, for the moment? But this is not for the upcoming release, yet, if I understood @joerick correctly.

Seems recent PRs resulted in conflicts, though. I'll rebase, in a bit!

Czaki commented:
@YannickJadoul I do not want to create next PR until types are finished/decided.

YannickJadoul commented:
Rebased

Czaki commented:
There is backport dataclasses module to python 3.6 https://pypi.org/project/dataclasses/

YannickJadoul commented:
> There is backport dataclasses module to python 3.6 https://pypi.org/project/dataclasses/

Is that worth it? Not a big advantage over `NamedTuple`, I thought?

Czaki commented:
dataclass allow to create modifiable container. 
It comes outside typing module: 

> I don't mind importing `typing` for type annotations, but it's weird/ugly that it suddenly takes over the role of `collections.namedtuple`. `typing` is for types, not for functionality!?

It also have few nice features like `dataclasses.field`

YannickJadoul commented:
> dataclass allow to create modifiable container.

But if we don't need mutable containers, it might be better to not do this?

> It comes outside typing module:
> 
> > I don't mind importing `typing` for type annotations, but it's weird/ugly that it suddenly takes over the role of `collections.namedtuple`. `typing` is for types, not for functionality!?
> 
> It also have few nice features like `dataclasses.field`

OK, yes, agreed there's some nice things, definitely. But not sure it's worth adding the extra dependency to the core of `cibuildwheel`? :-/

Czaki commented:
This is dependency only for python 3.6. From Python 3.7 it is built-in module. 

The question is If nice features of dataclass will be useful. If Yes then I not think it is good to wait with this till python 3.6 is dropped if it can be introduced earlier.

joerick commented:
I'm not so interested in the dataclass - config should be an immutable object anyway.

I think we should try to get this in soon, before #334, if that works for you @YannickJadoul! I'll take another look over now, but I think it's pretty close, it probably just needs a merge of upstream master.

joerick commented on cibuildwheel/linux.py:
```diff
@@ -42,8 +43,13 @@ def matches_platform(identifier):
+    path: str
+
+
+def get_python_configurations(build_selector: Callable[[str], bool]) -> List[PythonConfiguration]:
```
I think we'd be better to use BuildSelector as a type here - it'll help readability for people new to the codebase.

YannickJadoul replied:
OK, yeah, that might be a good argument. Doing so.

joerick commented on cibuildwheel/macos.py:
```diff
@@ -26,8 +28,13 @@ def call(args, env=None, cwd=None, shell=False):
+    url: str
+
+
+def get_python_configurations(build_selector: Callable[[str], bool]) -> List[PythonConfiguration]:
```
BuildSelector here?

YannickJadoul commented:
Alright. I'll see if I can fit in some time to rebase today or tomorrow. Still want to/need to find some time to decently review #334 as well.

joerick commented on cibuildwheel/windows.py:
```diff
@@ -3,37 +3,45 @@
+    url: Optional[str]
+
+
+def get_python_configurations(build_selector: Callable[[str], bool]) -> List[PythonConfiguration]:
```
BuildSelector?

YannickJadoul commented on CI.md:
```diff
@@ -1,11 +1,11 @@
+|          | 3.6                          | 3.7                                                | 3.8              |
+|----------|------------------------------|----------------------------------------------------|------------------|
+| Linux    | Travis CI / CircleCI         | AppVeyor / GitHub Actions                          | Azure Pipelines  |
+| macOS    | CircleCI                     | AppVeyor / Travis CI¹ / CircleCI / GitHub Actions  | Azure Pipelines  |
```
Not sure why that was originally, there, but yes. If it works for 3.6 on CircleCI, and works on other CIs for 3.7, then we should be safe!

YannickJadoul commented:
> Just a few notes. Then we'd need:

Thanks! Rebased, solved the conflicts, and fixed the final few things

> * some mypy config somewhere (equivalent to `--strict`), either in setup.cfg or in mypy.ini

Done. Added to `setup.cfg` since that already contains `flake8` config anyway. Or any reason to prefer `mypy.ini`?

There's [a bunch of other flags/config values to be set](https://mypy.readthedocs.io/en/stable/config_file.html), with various more errors/warnings. Nothing immediately catches my eye, but might be interesting to go over? Depends a bit on how strict you want to be; e.g. disallowing `Any` could be interesting in some places, to ensure annotations will be added in new code?

> * Type checking in CI. I guess either this is an explicit job, or maybe this would be convenient to roll it into flake8 https://pypi.org/project/flake8-mypy/

I was planning to just install `mypy` and run it in the same job (but different call) as `flake8`. Do you prefer `flake8-mypy`? To be fair, it's neat that it can be integrated, but I don't really see the advantage to add it as extra dependency (but it should be a minor/stable dependency, ofc).

YannickJadoul commented:
Hèhè, thanks! :-)
